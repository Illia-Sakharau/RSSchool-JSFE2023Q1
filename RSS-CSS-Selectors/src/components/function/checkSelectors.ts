export default function (html: string, checkedSelectors: string | null): boolean {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(html, 'text/xml');
  const correctAnswer = xmlDoc.querySelectorAll('[data-target="true"]');
  if (!checkedSelectors) {
    return false;
  }
  if (checkedSelectors === '[data-target="true"]') {
    alert(`Don't cheat, please!`);
    return false;
  }
  const enteredAnswer = xmlDoc.querySelectorAll(checkedSelectors);

  if (JSON.stringify(correctAnswer) !== JSON.stringify(enteredAnswer)) {
    return false;
  }

  return true;
}
