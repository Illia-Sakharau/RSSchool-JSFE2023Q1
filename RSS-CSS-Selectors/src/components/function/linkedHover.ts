export default (event: MouseEvent): void => {
  const foo =
    event.type === 'mouseout' ? (el: Element) => el.classList.remove('hovered') : (el: Element) => el.classList.add('hovered');

  const target = event.target;
  if (target instanceof HTMLElement) {
    const closest = target.closest('[data-object]');
    if (closest instanceof HTMLElement) {
      const object = closest.dataset.object;
      const num = closest.dataset.num;
      const num1 = closest.dataset.num1;
      const elements = document.querySelectorAll(`[data-object="${object}"][data-num="${num}"][data-num1="${num1}"]`);
      elements.forEach(foo);
    }
  }
};
