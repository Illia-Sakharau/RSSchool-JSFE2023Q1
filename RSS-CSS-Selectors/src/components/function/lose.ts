export default () => {
  console.log('lose');
};
