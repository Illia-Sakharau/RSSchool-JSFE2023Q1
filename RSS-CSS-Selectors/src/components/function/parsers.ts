import { Elements, IParsedElem, ParsedElementsArray } from '../../types/types';
import createElement from '../../utils/createElement';
import htmlToElement from '../../utils/htmlToElement';

export function parserMapToArray(map: string): ParsedElementsArray {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(map, 'text/xml');

  function travers(child: Element): ParsedElementsArray {
    if (child.children.length === 0) {
      const tag = child.tagName as Elements;
      const classes = [...child.classList];
      const target = child.attributes.getNamedItem('data-target') ? true : false;

      return [{ tag: tag, classes: classes, target: target }];
    }
    const arr: ParsedElementsArray = [];
    for (let i = 0; i < child.children.length; i++) {
      if (child.children[i].children.length !== 0) {
        const subArr: ParsedElementsArray = [];

        const tag = child.children[i].tagName as Elements;
        const classes = [...child.children[i].classList];
        const target = child.children[i].attributes.getNamedItem('data-target') ? true : false;

        subArr.push({ tag: tag, classes: classes, target: target });
        subArr.push(travers(child.children[i]));
        arr.push(subArr);
      } else {
        arr.push(...travers(child.children[i]));
      }
    }
    return arr;
  }

  return travers(xmlDoc.children[0]);
}

export function parserArrayToHTMLeditor(arrMap: ParsedElementsArray): HTMLElement {
  const prepareCode: HTMLElement = travers(arrMap);

  function createLine(isCloses: boolean, isBlock: boolean, obj: IParsedElem, padding: number): string {
    const { tag, classes } = obj;
    const tagStr = `${isCloses ? '/' : ''}<span class="tag">${tag}</span>`;
    const classesStr = classes?.length !== 0 && !isCloses ? `<span class="classes"> class="${classes}"</span>` : '';
    const oneTag = !isBlock && !isCloses ? ' /' : '';
    const template = `<div class="line" data-padding="${padding}">
        &lt;${tagStr}${classesStr}${oneTag}&gt;
      </div>
      `;
    return template;
  }

  function travers(obj: ParsedElementsArray, padding: number = 0, num: number = 0): HTMLElement {
    const res: HTMLElement = createElement({ tag: 'div' });
    if (!Array.isArray(obj)) {
      const elem = htmlToElement(createLine(false, false, obj, padding));
      res.append(elem);
      res.dataset.object = `${JSON.stringify(obj).split('"').join('')}`;
      res.dataset.num = `${num}`;
      return res;
    }
    if (!Array.isArray(obj[0]) && Array.isArray(obj[1])) {
      const first = htmlToElement(createLine(false, true, obj[0], padding));
      const last = htmlToElement(createLine(true, true, obj[0], padding));
      const inner = travers(obj[1] as ParsedElementsArray, padding + 1, num);
      res.append(first, inner, last);
      res.dataset.object = `${JSON.stringify(obj).split('"').join('')}`;
      res.dataset.num = `${num}`;
      return res;
    }
    obj.forEach((el, i) => {
      const numEl = num || i;
      const elem = travers(el as ParsedElementsArray, padding + 1, numEl);
      elem.dataset.num = `${numEl}`;
      elem.dataset.num1 = `${i}`;
      res.append(elem);
    });
    return res;
  }
  prepareCode.prepend(htmlToElement(`<div class="line">&lt;shelf&gt</div>`));
  prepareCode.append(htmlToElement(`<div class="line">&lt;/shelf&gt</div>`));
  return prepareCode;
}
