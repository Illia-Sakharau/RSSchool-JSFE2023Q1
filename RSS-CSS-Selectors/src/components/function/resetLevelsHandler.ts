import { LEVELS_STATES } from '../../data/constants';

export default () => {
  localStorage.clear();
  LEVELS_STATES.fill(null);
  const levelChanged = new Event('levelChanged');
  document.dispatchEvent(levelChanged);
};
