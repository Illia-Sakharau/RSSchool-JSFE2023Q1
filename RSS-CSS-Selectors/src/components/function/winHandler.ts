export default () => {
  alert(`Congratulations!!! You completed all levels!`);
};
