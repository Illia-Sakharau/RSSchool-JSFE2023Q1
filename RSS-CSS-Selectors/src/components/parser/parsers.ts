import { Elements, IParsedElem, ParsedElementsArray } from '../../types/types';
import createElement from '../../utils/createElement';
import htmlToElement from '../../utils/htmlToElement';

export function parserMapToArray(map: string): ParsedElementsArray {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(map, 'text/xml');

  function travers(child: Element): ParsedElementsArray {
    if (child.children.length === 0) {
      const tag = child.tagName as Elements;
      const classes = [...child.classList];
      const target = child.attributes.getNamedItem('data-target') ? true : false;

      return [{ tag: tag, classes: classes, target: target }];
    }
    const arr: ParsedElementsArray = [];
    for (let i = 0; i < child.children.length; i++) {
      if (child.children[i].children.length !== 0) {
        const subArr: ParsedElementsArray = [];

        const tag = child.children[i].tagName as Elements;
        const classes = [...child.children[i].classList];
        const target = child.children[i].attributes.getNamedItem('data-target') ? true : false;

        subArr.push({ tag: tag, classes: classes, target: target });
        subArr.push(travers(child.children[i]));
        arr.push(subArr);
      } else {
        arr.push(...travers(child.children[i]));
      }
    }
    return arr;
  }

  return travers(xmlDoc.children[0]);
}

export function parserArrayToHTMLeditor(arrMap: ParsedElementsArray): { prepareCode: HTMLElement; elemArr: HTMLElement[] } {
  const elemArr: HTMLElement[] = [];
  console.error('удалить массив если он не понадабится');
  const prepareCode: HTMLElement = travers(arrMap);

  function createLine(isCloses: boolean, isBlock: boolean, obj: IParsedElem, padding: number): string {
    const { tag, classes } = obj;
    const tagStr = `${isCloses ? '/' : ''}<span class="tag">${tag}</span>`;
    const classesStr = classes?.length === 0 ? '' : `<span class="classes"> class="${classes}"</span>`;
    const oneTag = !isBlock && !isCloses ? ' /' : '';
    const template = `<div class="line" data-padding="${padding}">
        &lt;${tagStr}${classesStr}${oneTag}&gt;
      </div>
      `;
    return template;
  }

  function travers(obj: ParsedElementsArray, padding: number = 0): HTMLElement {
    const res: HTMLElement = createElement({ tag: 'div' });
    if (!Array.isArray(obj)) {
      const { tag } = obj;
      const elem = htmlToElement(createLine(false, false, obj, padding));
      res.append(elem);
      res.dataset.tagName = `${tag}`;
      res.dataset.colNum = `${padding}`;
      return res;
    }
    if (!Array.isArray(obj[0]) && Array.isArray(obj[1])) {
      const { tag } = obj[0];
      const first = htmlToElement(createLine(false, true, obj[0], padding));
      const last = htmlToElement(createLine(true, true, obj[0], padding));
      const inner = travers(obj[1] as ParsedElementsArray, padding + 1);
      res.append(first, inner, last);
      res.dataset.tagName = `${tag}`;
      res.dataset.colNum = `${padding}`;
      return res;
    }
    obj.forEach((el, i) => {
      const elem = travers(el as ParsedElementsArray, padding + 1);
      elem.dataset.rowNum = `${i}`;
      elemArr.push(elem);
      res.append(elem);
    });
    return res;
  }
  return { prepareCode, elemArr };
}
