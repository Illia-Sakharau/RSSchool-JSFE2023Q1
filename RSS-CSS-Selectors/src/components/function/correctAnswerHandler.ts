import { CURRENT_LEVEL, LEVELS_STATES } from '../../data/constants';
import winHandler from './winHandler';

export default () => {
  const targetsElems = document.querySelectorAll('.target');
  const currentLevel = CURRENT_LEVEL.getLevel();
  let nextLvl = currentLevel < LEVELS_STATES.length - 1 ? currentLevel + 1 : 0;
  let isWin = false;

  if (!LEVELS_STATES[currentLevel]) {
    LEVELS_STATES[currentLevel] = 'completed';
    localStorage.setItem(`ily-level${currentLevel}`, 'completed');
  }

  if (LEVELS_STATES.every((lvl) => !!lvl)) {
    isWin = true;
  } else {
    while (LEVELS_STATES[nextLvl] !== null) {
      nextLvl = nextLvl < LEVELS_STATES.length - 1 ? nextLvl + 1 : 0;
    }
  }

  CURRENT_LEVEL.setCurentLvl(nextLvl);
  localStorage.setItem('ily-currentLvl', `${nextLvl}`);

  targetsElems.forEach((el) => el.classList.add('correctAnswer'));
  targetsElems[0].addEventListener(
    'animationend',
    () => {
      if (isWin) winHandler();
      const levelChanged = new Event('levelChanged');
      document.dispatchEvent(levelChanged);
      return;
    },
    false,
  );
};
