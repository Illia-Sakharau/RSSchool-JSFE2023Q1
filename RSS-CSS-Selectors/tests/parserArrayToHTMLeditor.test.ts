import { parserArrayToHTMLeditor } from '../src/components/function/parsers';
import { currentLevel } from './_test-data';

const testArrayMap = currentLevel.map;

let createHTMLViewerCode: HTMLElement;
beforeEach(() => {
  createHTMLViewerCode = parserArrayToHTMLeditor(testArrayMap);
});

describe('Test parser array map to html element viewer HTML code', () => {
  test('should be defined', () => {
    expect(createHTMLViewerCode).toBeDefined();
  });
  test('should return value instance of HTMLElement', () => {
    expect(createHTMLViewerCode).toBeInstanceOf(HTMLElement);
  });
  test('should return correct answer', () => {
    expect(createHTMLViewerCode).toMatchSnapshot();
  });
});
