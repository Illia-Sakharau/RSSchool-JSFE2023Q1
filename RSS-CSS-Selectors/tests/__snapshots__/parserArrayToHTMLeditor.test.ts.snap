// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test parser array map to html element viewer HTML code should return correct answer 1`] = `
<div>
  <div
    class="line"
  >
    &lt;shelf&gt;
  </div>
  <div
    data-num="1994"
    data-num1="0"
    data-object="{tag:rack,classes:[],target:false,id:}"
  >
    <div
      class="line"
      data-padding="1"
    >
      
        &lt;
      <span
        class="tag"
      >
        rack
      </span>
       /&gt;
      
    </div>
  </div>
  <div
    data-num="1995"
    data-num1="1"
    data-object="{tag:pot,classes:[],target:true,id:}"
  >
    <div
      class="line"
      data-padding="1"
    >
      
        &lt;
      <span
        class="tag"
      >
        pot
      </span>
       /&gt;
      
    </div>
  </div>
  <div
    data-num="1996"
    data-num1="2"
    data-object="{tag:chamomile,classes:[],target:false,id:}"
  >
    <div
      class="line"
      data-padding="1"
    >
      
        &lt;
      <span
        class="tag"
      >
        chamomile
      </span>
       /&gt;
      
    </div>
  </div>
  <div
    data-num="1997"
    data-num1="3"
    data-object="{tag:tulip,classes:[],target:false,id:}"
  >
    <div
      class="line"
      data-padding="1"
    >
      
        &lt;
      <span
        class="tag"
      >
        tulip
      </span>
       /&gt;
      
    </div>
  </div>
  <div
    class="line"
  >
    &lt;/shelf&gt;
  </div>
</div>
`;
